crates/cadenza/src/mcp.rs

```diff
@@ -0,0 +1,430 @@
+        let mut output = String::new();
+        if !parsed.errors.is_empty() {
+            output.push_str("Errors:\n");
+            for err in &parsed.errors {
+                output.push_str(&format!("  {}\n", err.message));
+            }
+            output.push('\n');
+        }
+
+        // Format AST
+        output.push_str("AST:\n");
+        output.push_str(&format_syntax_tree(&parsed.syntax(), 0));
+
+        Ok(CallToolResult::success(vec![Content::text(output)]))
+    }
+
+    /// Infer the type of a Cadenza expression
+    #[tool(description = "Infer the type of a Cadenza expression. Use this to understand what type an expression evaluates to.")]
+    async fn infer_type(
+        &self,
+        rmcp::handler::server::wrapper::Parameters(req): rmcp::handler::server::wrapper::Parameters<
+            InferTypeRequest,
+        >,
+    ) -> Result<CallToolResult, McpError> {
+        let mut compiler = Compiler::new();
+        let mut env = Env::new();
+
+        let parsed = parse(&req.expression);
+        if !parsed.errors.is_empty() {
+            return Ok(CallToolResult::error(vec![Content::text(
+                "Parse errors encountered".to_string(),
+            )]));
+        }
+
+        // Evaluate to get the type
+        let result = cadenza_eval::eval(&parsed.ast(), &mut env, &mut compiler);
+        // eval returns Vec<Value>, we'll check the type of the last value
+        let type_str = if let Some(value) = result.last() {
+            match value {
+                Value::Nil => "Nil".to_string(),
+                Value::Bool(_) => "Bool".to_string(),
+                Value::Symbol(_) => "Symbol".to_string(),
+                Value::Integer(_) => "Integer".to_string(),
+                Value::Float(_) => "Float".to_string(),
+                Value::String(_) => "String".to_string(),
+                Value::List(_) => "List".to_string(),
+                Value::Record { .. } => "Record".to_string(),
+                Value::StructConstructor { name, .. } => format!("StructConstructor<{}>", name),
+                Value::Type(_) => "Type".to_string(),
+                Value::Quantity { dimension, .. } => format!("Quantity<{}>", dimension),
+                Value::UnitConstructor(_) => "UnitConstructor".to_string(),
+                Value::BuiltinFn(_) => "BuiltinFn".to_string(),
+                Value::BuiltinMacro(_) => "BuiltinMacro".to_string(),
+                Value::SpecialForm(_) => "SpecialForm".to_string(),
+                Value::UserFunction(_) => "UserFunction".to_string(),
+            }
+        } else {
+            "Nil".to_string()
+        };
+        Ok(CallToolResult::success(vec![Content::text(format!(
+            "Type: {}",
+            type_str
+        ))]))
+    }
+
+    /// Check dimensional analysis for an expression with units
+    #[tool(description = "Check dimensional analysis for an expression with units. Use this to verify unit compatibility and conversions.")]
```

Comment:
Isn't this just an interpreter? Do we need two commands that do the exact same thing?